Associated type bounded on another associated type with a default fails

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#![feature(associated_type_defaults)]

use std::io::Read;

trait View {
    type Deserializers: Deserializer<Item = Self::RequestParams>;
    type RequestParams = DefaultRequestParams;
}

struct DefaultRequestParams;

trait Deserializer {
    type Item;
    fn deserialize(r: impl Read) -> Self::Item;
}

```

I expected to see this happen: `View::Deserializers` has the as of yet unspecified type `Deserializer<Item=Self::RequestParams>`.

Instead, this happened: 

```rust
error[E0271]: type mismatch resolving `<<Self as View>::Deserializers as Deserializer>::Item == DefaultRequestParams`
  --> src\lib.rs:17:38
   |
14 | trait View {
   | -------------- required by `View`
...
17 |     type Deserializers: Deserializer<Item = Self::RequestParams>;
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `DefaultRequestParams`, found associated type
   |
   = note:       expected struct `DefaultRequestParams`
           found associated type `<Self as View>::RequestParams`
   = help: consider constraining the associated type `<Self as View>::RequestParams` to `DefaultRequestParams`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

```

This error does not occur when the default is removed.

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.47.0-nightly (f44c6e4e2 2020-08-24)
```

---

See also: [My initial question on the internals forum](https://internals.rust-lang.org/t/error-in-interaction-between-associated-type-defaults-and-associated-type-bounds/12978)

[RFC 2532 (assocaited type defaults](https://github.com/rust-lang/rust/issues/29661)

cc @jonas-schievink

--- a/src/test/ui/issues/issue-75763.rs
+++ b/src/test/ui/issues/issue-75763.rs
@@ -8,6 +8,7 @@ struct Bug<const S: &'static str>;
 fn main() {
     let b: Bug::<{
         unsafe {
+            // FIXME(const_generics): Decide on how to deal with invalid values as const params.
             std::mem::transmute::<&[u8], &str>(&[0xC0, 0xC1, 0xF5])
         }
     }>;

--- a/src/test/ui/issues/issue-75763.rs
+++ b/src/test/ui/issues/issue-75763.rs
@@ -1,0 +1,14 @@
+// build-pass
+
+#![allow(incomplete_features)]
+#![feature(const_generics)]
+
+struct Bug<const S: &'static str>;
+
+fn main() {
+    let b: Bug::<{
+        unsafe {
+            std::mem::transmute::<&[u8], &str>(&[0xC0, 0xC1, 0xF5])
+        }
+    }>;
+}

--- a/src/test/ui/issues/issue-76179.rs
+++ b/src/test/ui/issues/issue-76179.rs
@@ -1,0 +1,19 @@
+// check-pass
+
+#![feature(associated_type_defaults)]
+
+use std::io::Read;
+
+trait View {
+    type Deserializers: Deserializer<Item = Self::RequestParams>;
+    type RequestParams = DefaultRequestParams;
+}
+
+struct DefaultRequestParams;
+
+trait Deserializer {
+    type Item;
+    fn deserialize(r: impl Read) -> Self::Item;
+}
+
+fn main() {}

